vncviewer/EmulateMB.h

```diff
@@ -21,29 +21,50 @@
 
 #include <rfb/Timer.h>
 #include <rfb/Rect.h>
+#include <rdr/types.h>
 
 class EmulateMB : public rfb::Timer::Callback {
 public:
-  EmulateMB();
+  EmulateMB(rdr::U32 emulateMBModKey);
 
   void filterPointerEvent(const rfb::Point& pos, int buttonMask);
 
+  bool filterKeyPress(int keyCode, rdr::U32 keySym);
+  bool filterKeyRelease(int keyCode);
+
 protected:
   virtual void sendPointerEvent(const rfb::Point& pos, int buttonMask)=0;
 
   virtual bool handleTimeout(rfb::Timer *t);
 
+  virtual void writeKeyEvent(rdr::U32 keySym, rdr::U32 keyCode, bool down)=0;
+
 private:
+  typedef const signed char Action[3];
+  static Action mbStateTab[8][5];
   void sendAction(const rfb::Point& pos, int buttonMask, int action);
 
   int createButtonMask(int buttonMask);
 
+  void mbFsmDoMod(int modAction);
+  void mbFsmDoButton(const rfb::Point& pos, int buttonMask, int buttonAction);
+  bool filterKeyPressRelease(bool is_press, int keyCode, rdr::U32 keySym);
+  void mbFSM(Action action);
+  void filterPointerEventMod(const rfb::Point& pos, int buttonMask);
+  void filterPointerEventLR(const rfb::Point& pos, int buttonMask);
+
 private:
   int state;
   int emulatedButtonMask;
   int lastButtonMask;
   rfb::Point lastPos, origPos;
   rfb::Timer timer;
+
+  // Variables for mod-button1 emulation
+  rdr::U32 emulateMiddleButtonModifierKey;	// The key sym used for emulation
+  int mbState;
```

Comment:
Can't we use the existing `state`? Both modes cannot be active at the same time.